controller/api/destination/endpoint_translator.go

```diff
@@ -402,6 +403,9 @@ func createWeightedAddr(address watcher.Address, opaquePorts map[uint32]struct{}
 				weightedAddr.ProtocolHint.OpaqueTransport = &pb.ProtocolHint_OpaqueTransport{
 					InboundPort: port,
 				}
+				weightedAddr.ProtocolHint.Protocol = &pb.ProtocolHint_Opaque_{
+					Opaque: &pb.ProtocolHint_Opaque{},
+				}
 			}
 		}
```

Comment:
I think this might read more clearly if we put the `enableH2Upgrade` block as an `else if` here.  This helps communicate that these are mutually exclusive and that opaque takes precedence, rather than relying on ordering and overwriting.  However, if we do that, we'll need to be a bit careful to make sure we properly handle the case when OpaqueProtocol is set but `getInboundPort` fails (what even IS the correct behavior in that case?).

Reply:
hmm, AFAICT if `getInboundPort` fails, that means that we couldn't parse a port from the pod spec's proxy container's env vars, or the pod doesn't have a proxy container. if there's no proxy at all, we shouldn't be enabling the opaque transport flag... but if the env var is malformed, what should we do? just skip the update entirely?

Reply:
Previously, if `getInboundPort` failed, we would log and leave the protocol hint set to H2 (if enabled).  Not sure if that was the right behavior to begin with, but after this change the behavior would be to leave the protocol completely unset.

Would it make sense to still set ProtocolHint_Opaque, even if the getInboundPort fails?  Is that a valid configuration? Can opaque traffic be sent without the tagged transport?

Reply:
> Previously, if `getInboundPort` failed, we would log and leave the protocol hint set to H2 (if enabled). Not sure if that was the right behavior to begin with, but after this change the behavior would be to leave the protocol completely unset.

I think the new behavior is more correct. We should not be attempting an H2 upgrade if targeting a port that's marked as opaque, since we know that the proxy on the receiving end will not attempt a downgrade back to HTTP/1.

> Would it make sense to still set ProtocolHint_Opaque, even if the getInboundPort fails? Is that a valid configuration? Can opaque traffic be sent without the tagged transport?

The `Opaque` protocol hint is meaningless without tagged transport, as it only configures what `SessionProtocol` is sent when using tagged transport. perhaps we should still send it in this case, but we should _not_ be populating the `ProtocolHint.OpaqueTransport` field, as we don't know the port that connections should be directed to, and therefore cannot use opaque transport.